tests/Compat/PolyfillAssertTrait.php

```diff
@@ -0,0 +1,25 @@
+<?php
+
+namespace MongoDB\Tests\Compat;
+
+use PHPUnit\Framework\Assert;
+use ReflectionClass;
+use Symfony\Bridge\PhpUnit\Legacy\PolyfillAssertTrait as SymfonyPolyfillAssertTrait;
+
+$r = new ReflectionClass(Assert::class);
+if (! $r->hasMethod('assertEqualsWithDelta')) {
+    /**
+     * @internal
+     */
+    trait PolyfillAssertTrait
+    {
+        use SymfonyPolyfillAssertTrait;
```

Comment:
Is there a particular reason that [`PolyfillAssertTrait`](https://github.com/symfony/symfony/blob/4.4/src/Symfony/Bridge/PhpUnit/Legacy/PolyfillAssertTrait.php) doesn't have a compat shim, like [`SetUpTearDownTrait`](https://github.com/symfony/symfony/blob/4.4/src/Symfony/Bridge/PhpUnit/SetUpTearDownTrait.php)?

Is that because it really is only relevant to older PHPUnit versions, so Symfony is careful not to use it in newer branches?

Reply:
I talked to @nicolas-grekas about this. The assertion polypill is marked as `@internal` because they only use it to patch PHPUnit classes in their simple-phpunit binary: https://github.com/symfony/symfony/blob/a9ace363899373f4fed96f32cd38fc94187666d2/src/Symfony/Bridge/PhpUnit/bin/simple-phpunit.php#L154..L157. With the myriad of PHPUnit versions Symfony needs to support, it's too risky for them to have a single trait and let people use that in classes extending from `PHPUnit\Framework\Assert`, as it would then override the methods already shipped by PHPUnit. Patching the original class works around this issue as the code in `PHPUnit\Framework\Assert` then overrides the trait code.

In our case, I didn't want to patch PHPUnit code and with the PHPUnit versions we currently support, the single trait is good enough for us as all new assertion methods were added in or after PHPUnit 7. That's why we declare our own trait, which is empty if the PHPUnit version is new enough, as it allows us to always include our polyfill trait in the test classes.

There is still a certain risk using internal code from Symfony, but given that they need to keep the trait around as long as they support PHPUnit < 7 I feel rather confident there: dropping older PHPUnit versions for them won't be possible until Symfony 4.4 is EOL, which will be in November 2023. I hope we'll be able to get rid of support for PHP 5 before then. However, if you'd rather not use an internal class we can always duplicate the polyfill to our library, of course keeping the original copyright intact.

Reply:
Or you could use simple-phpunit :)
But I agree, the risk is low.

Reply:
Thanks for the deep explanation! I'm content with this.

@nicolas-grekas: Thanks for chiming in. `simple-phpunit` looks interesting (first I've seen of it) and is probably something (along with the bridge) that I could have benefited from a good while back :smile:. I'll defer to @alcaeus about whether it's worth migrating over to that at some later point.